torchtune/data/_collate.py

```diff
@@ -396,102 +396,143 @@ def padded_collate_tiled_images_and_mask(
             )
         }
 
+    batch_dict = {
+        "tokens": collated_text["tokens"],
+    }
+    if "labels" in collated_text:
+        batch_dict["labels"] = collated_text["labels"]
+
     max_seq_len = collated_text["tokens"].shape[-1]
     bsz = len(batch)
 
     # TODO: Figure out how to make this more efficient or vectorized. Setting
     # max_num_tiles beforehand will save one nested for loop but may incur more
     # memory and compute costs in attention if max_num_tiles > batch_max_num_tiles
+    if (
+        "vision" in batch[0]["encoder_input"]
+        and len(batch[0]["encoder_input"]["vision"]["images"]) > 0
```

Comment:
why is "vision" sometimes nested?

Reply:
didn't you write this code lol

Reply:
Anyways this is cause of the Llama4 transform. Now we have "encoder_input": {"vision": {...}, ...} for ... reasons... Strictly speaking we probably don't need it, but it does allow us to generalize to other modalities if we need